lightning/src/ln/channelmanager.rs

```diff
@@ -4509,12 +4568,91 @@ impl_writeable_tlv_based!(HTLCPreviousHopData, {
 	(6, incoming_packet_shared_secret, required)
 });
 
-impl_writeable_tlv_based!(ClaimableHTLC, {
-	(0, prev_hop, required),
-	(2, value, required),
-	(4, payment_data, required),
-	(6, cltv_expiry, required),
-});
+impl Writeable for ClaimableHTLC {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		let payment_data = match &self.claim_type {
+			ClaimableType::Invoice(data) => data.clone(),
+			_ => msgs::FinalOnionHopData {
+				payment_secret: PaymentSecret([0; 32]),
+				total_msat: 0,
+			}
+		};
+		let keysend_preimage = match self.claim_type {
+			ClaimableType::Invoice(_) => None,
+			ClaimableType::Spontaneous(preimage) => Some(preimage.clone()),
+		};
+		write_tlv_fields!
+		(writer,
+		 {
+		   (0, self.prev_hop, required), (2, self.value, required),
+		   (4, payment_data, required), (6, self.cltv_expiry, required),
+			 (8, keysend_preimage, option),
+		 });
+		Ok(())
+	}
```

Comment:
Do we need to do this because we introduced an enum where there wasn't one before? Would it make sense to bump the version and serialize enums using the macro for doing so?

Reply:
Yeah: https://github.com/rust-bitcoin/rust-lightning/pull/967#discussion_r663983938 

> Would it make sense to bump the version and serialize enums using the macro for doing so?

Is this the same idea as in [this comment](https://github.com/rust-bitcoin/rust-lightning/pull/967#discussion_r668378341)? I'm pretty eager to get this PR out the door to unblock sphinx, but would love to get @TheBlueMatt's input on this idea since it does sound potentially nicer

Reply:
Yeah, similar concern.

Reply:
We can't "just bump the version" since we want to at least *try* (I think) to be compatible where old versions can read new serialized data. Because some users will be synchronizing state across multiple applications that may upgrade at different times, I think we should generally try to at least have our data be readable by the previous version as long as you aren't using new features.